File: server/src/main/java/model/loot/LootTable.java
package model.loot;

import manager.Database;
import manager.LanguageUtility;
import manager.Utility;
import model.ICurrency;
import model.item.IItem;
import model.item.Item;
import org.apache.commons.configuration2.Configuration;

import java.util.ArrayList;
import java.util.Collection;

public class LootTable implements ILootTable {

    private ArrayList<ILootFactory> list;

    public LootTable() {
        this.list = new ArrayList<>();
    }

    public void add(String name, int amount, double chance) {
        list.add(new LootFactory(Database.getItem(name), amount, chance));
    }

    public void add(ILootFactory factory) {
        list.add(factory);
    }

    public void add(ICurrency currency) {
        Configuration config = Utility.getConfig();
        int silverToCopper = config.getInt("coin.silver.toCopper");
        int goldToCopper = config.getInt("coin.gold.toSilver") * silverToCopper;

        int value = currency.getCoinValue();
        int gp = value / goldToCopper;
        value -= gp * goldToCopper;

        int sp = value / silverToCopper;
        value -= sp * silverToCopper;

        int cp = value;

        Item gold = Database.getItem(LanguageUtility.getMessage("coin.gold"));
        Item silver = Database.getItem(LanguageUtility.getMessage("coin.silver"));
        Item copper = Database.getItem(LanguageUtility.getMessage("coin.copper"));

        add(gold, gp, 1);
        add(silver, sp, 1);
        add(copper, cp, 1);
    }

    public void add(IItem item, int amount, double chance) {
        list.add(new LootFactory(item, amount, chance));
    }

    public void add(ILootTable other) {
        this.list.addAll(other.getLootFactories());
    }

    @Override
    public Collection<ILootFactory> getLootFactories() {
        return list;
    }

    @Override
    public void setLootFactories(Collection<ILootFactory> factories) {
        this.list = new ArrayList<>(factories);
    }

    @Override
    public Collection<ILoot> getLoot() {

        ArrayList<ILoot> lootList = new ArrayList<>();

        for (ILootFactory factory : list) {
            ILoot loot = factory.getLoot();
            ILoot own = getLoot(factory.getItem(), lootList);

            if (own != null) {
                own.addAmount(loot.getAmount());
            } else {
                if (loot.getAmount() > 0) {
                    lootList.add(loot);
                }
            }
        }

        return lootList;
    }

    private ILoot getLoot(IItem artifact, ArrayList<ILoot> list) {
        for (ILoot l : list) {
            if (l.getItem().equals(artifact)) {
                return l;
            }
        }
        return null;
    }
}

File: app/src/main/java/de/pnp/manager/app/model/LootTable.kt
package de.pnp.manager.app.model

import de.pnp.manager.app.state.ConfigHelper
import de.pnp.manager.app.state.DatabaseHelper
import de.pnp.manager.model.ICurrency
import de.pnp.manager.model.item.IItem
import de.pnp.manager.model.loot.ILoot
import de.pnp.manager.model.loot.ILootFactory
import de.pnp.manager.model.loot.ILootTable

class LootTable : ILootTable {
    private var list: ArrayList<ILootFactory>? = null

    override fun add(name: String, amount: Int, chance: Double) {
        list!!.add(
            LootFactory(
                DatabaseHelper.getItem(name),
                amount,
                chance
            )
        )
    }

    override fun add(factory: ILootFactory?) {
        list!!.add(factory!!)
    }

    override fun add(currency: ICurrency?) {
        val silverToCopper: Int = ConfigHelper.getInt("coin.silver.toCopper")
        val goldToCopper: Int = ConfigHelper.getInt("coin.gold.toSilver") * silverToCopper

        var value = currency!!.coinValue
        val gp = value / goldToCopper
        value -= gp * goldToCopper

        val sp = value / silverToCopper
        value -= sp * silverToCopper

        val cp = value

        val gold: IItem =
            DatabaseHelper.getItem("Gold")
        val silver: IItem =
            DatabaseHelper.getItem("Silber")
        val copper: IItem =
            DatabaseHelper.getItem("Kupfer")

        add(gold, gp, 1.0)
        add(silver, sp, 1.0)
        add(copper, cp, 1.0)
    }

    override fun add(item: IItem?, amount: Int, chance: Double) {
        list!!.add(LootFactory(item, amount, chance))
    }

    override fun add(other: ILootTable?) {
        list!!.addAll(other!!.lootFactories)
    }

    override fun getLootFactories(): MutableCollection<ILootFactory> {
        return list!!
    }

    override fun setLootFactories(factories: MutableCollection<ILootFactory>?) {
        this.list = ArrayList(factories)
    }

    override fun getLoot(): MutableCollection<ILoot> {
        val lootList = java.util.ArrayList<ILoot>()

        for (factory in list!!) {
            val loot = factory.loot
            val own: ILoot? = getLoot(factory.item, lootList)
            if (own != null) {
                own.addAmount(loot.amount)
            } else {
                if (loot.amount > 0) {
                    lootList.add(loot)
                }
            }
        }

        return lootList
    }

    private fun getLoot(artifact: IItem, list: java.util.ArrayList<ILoot>): ILoot? {
        for (l in list) {
            if (l.item == artifact) {
                return l
            }
        }
        return null
    }
}